### PR TITLE
Init/v0.0.1

### PR DESCRIPTION
**Pull request by gagantrivedi**

### CODE REVIEW

File: flagsmith/provider_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	}
+}
+func testAccPreCheck(t *testing.T) {
+	mustHaveEnv(t, "FLAGSMITH_MASTER_API_KEY")
+	mustHaveEnv(t, "FLAGSMITH_FEATURE_NAME")
+	mustHaveEnv(t, "FLAGSMITH_ENVIRONMENT_KEY")
+	mustHaveEnv(t, "FLAGSMITH_ENVIRONMENT_ID")
+	mustHaveEnv(t, "FLAGSMITH_FEATURE_ID")
+}
+
```
Comment (matthewelwell) [inline flagsmith/provider_test.go:29]: Is it intentional to have blank space between certain functions and not others?

File: flagsmith/resource_flag.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	resp.Diagnostics.Append(diags...)
+	// Early return if the state is wrong
+	if diags.HasError() {
+		return
+	}
+	featureState, err := r.provider.client.GetEnvironmentFeatureState(data.EnvironmentKey.Value, data.FeatureName.Value)
+	if err != nil {
+		panic(err)
+	}
+	resoureData := MakeFlagResourceDataFromClientFS(featureState)
```
Comment (matthewelwell) [inline flagsmith/resource_flag.go:159]:
```suggestion
	resourceData := MakeFlagResourceDataFromClientFS(featureState)
```

File: flagsmith/resource_flag.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	elog := fmt.Sprintf("%+v", resoureData.FeatureStateValue)
+	tflog.Debug(ctx, elog)
+	diags = resp.State.Set(ctx, &resoureData)
+	if diags.HasError() {
+		// Log error from diags
+		for _, diag := range diags {
+			tflog.Error(ctx, diag.Detail())
+		}
+		resp.Diagnostics.Append(diags...)
+		tflog.Error(ctx, "Read: Error setting resource state")
```
Comment (matthewelwell) [inline flagsmith/resource_flag.go:171]: Not sure why the error message here is `"Read: Error setting resource state"` ? 'Read' seems to be irrelevant here?

File: flagsmith/resource_flag.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	intFeatureStateID, _ := state.ID.Value.Int64()
+	clientFeatureState := plan.ToClientFS(intFeatureStateID)
+
+	updatedClientFS, err := r.provider.client.UpdateFeatureState(clientFeatureState)
+
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update feature state, got error: %s", err))
+		return
+	}
+	resoureData := MakeFlagResourceDataFromClientFS(updatedClientFS)
```
Comment (matthewelwell) [inline flagsmith/resource_flag.go:205]:
```suggestion
	resourceData := MakeFlagResourceDataFromClientFS(updatedClientFS)
```

File: flagsmith/resource_flag.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+func (r flagResource) Delete(ctx context.Context, req tfsdk.DeleteResourceRequest, resp *tfsdk.DeleteResourceResponse) {
+	// Since deleting a feature state does not make sense, we do nothing
+	resp.State.RemoveResource(ctx)
+	return
+
+}
+
+func (r flagResource) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
+	idParts := strings.Split(req.ID, ",")
```
Comment (matthewelwell) [inline flagsmith/resource_flag.go:222]: I'm quite unclear what this code is doing? What are the `idParts` ?

File: docs/index.md [hunk 6]
```diff
@@ -23,4 +23,5 @@ provider "scaffolding" {
 
 ### Optional
 
-- `example` (String) Example provider attribute
+- `base_api_url` (String) Used by api client to connect to flagsmith instance. NOTE: update this if you are running a self hosted version
+- `master_api_key` (String, Sensitive) Master API key used by flagsmith api client. Can also be set using the environment variable `FLAGSMITH_MASTER_API_KEY`
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply docs/index.md:27]: That's actually  a common pattern to support passing `FLAGSMITH_MASTER_API_KEY` from the environment variable, because it's sensitive

File: flagsmith/resource_flag_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+					resource.TestCheckResourceAttr("flagsmith_flag.test_feature", "environment", strconv.Itoa(environmentID())),
+					resource.TestCheckResourceAttr("flagsmith_flag.test_feature", "feature", strconv.Itoa(featureID())),
+					resource.TestCheckResourceAttr("flagsmith_flag.test_feature", "feature_state_value.string_value", "one"),
+					resource.TestCheckResourceAttr("flagsmith_flag.test_feature", "enabled", "true"),
+				),
+			},
+			// Update testing
+
+			{
+				Config: testAccFlagResourceConfig("two", false),
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply flagsmith/resource_flag_test.go:46]: Yeah, it actually disables the feature and updates the feature value to `two`

File: docs/resources/flagsmith_flag.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+# flagsmith_flag (Resource)
+
+Flagsmith feature/ Remote config associated with an environment
+
+## Example Usage
+
+```terraform
+resource "flagsmith_flag" "feature_1_dev" {
+  enabled         = true
+  environment     = 2
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply docs/resources/flagsmith_flag.md]: Yeah, good point it's computed now

File: README.md [hunk 9]
```diff
@@ -43,10 +37,6 @@ go mod tidy
 
 Then commit the changes to `go.mod` and `go.sum`.
 
-## Using the provider
-
-Fill this in for each provider
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply README.md:48]: That's part of examples and docs, and it does not make sense to copy and paste that here from examples

File: flagsmith/resource_flag.go (hunk 3 above)
Comment (gagantrivedi) [reply flagsmith/resource_flag.go]: It's a common pattern in go log the error with function/mod name

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
 
-# This GitHub action runs your tests for each pull request and push.
-# Optionally, you can turn it on using a schedule for regular testing.
 on:
   pull_request:
     paths-ignore:
       - 'README.md'
   push:
+    branches:
+      - main
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply .github/workflows/test.yml:10]: Nah, I don't think so

File: docs/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply docs/index.md:28]: Well, because it is optional, else there is no way for us to read it from environment. If it's required, terraform will make sure it's in `.tf` hence optional

File: flagsmith/provider_test.go (hunk 1 above)
Comment (gagantrivedi) [reply flagsmith/provider_test.go:24]: Well, yes to sort of group them

File: flagsmith/provider_test.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/Flagsmith/flagsmith-go-api-client"
+	"github.com/Flagsmith/terraform-provider-flagsmith/flagsmith"
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"os"
+	"strconv"
+	"testing"
+)
+
+var testAccProtoV6ProviderFactories = map[string]func() (tfprotov6.ProviderServer, error){
```
[the remark this answers is not in the document]
Comment (gagantrivedi) [reply flagsmith/provider_test.go:13]: Yep, that's what the convention is since it's independent of a resource

**Review by matthewelwell — APPROVED** [review]
